### ISSUE TITLE
Preview feature feedback: `jsonProtocol`

### ISSUE THREAD

**Issue by janpio** [prisma/prisma#18095]
Our new Json Protocol is in early preview since 4.11.0.

During performance investigations and optimizations, we noticed that the existing protocol implementation using GraphQL added a CPU and memory overhead that was especially noticeable for larger Prisma schemas. Therefore, we found an alternative way to express our queries without needing that overhead: JSON.

To try out the new protocol, enable the `jsonProtocol` Preview feature in your Prisma schema:

```prisma
generator client {
  provider        = "prisma-client-js"  
  previewFeatures = ["jsonProtocol"]
}
```

Regenerate Prisma Client to use the new JSON protocol.

For environments or situations where it is not viable to enable the Preview feature flag to your Prisma schema file, we also added an environment variable that you can use to force the use of the JSON Protocol Preview feature: `PRISMA_ENGINE_PROTOCOL=json`.

~~Note: This is an early Preview feature with a significant limitation: Invalid input to Prisma Client will throw unpolished, internal errors that are less descriptive and user-friendly than our usual ones. We intend to improve these future releases.~~ ~~Using it with Data Proxy and Prisma Data Platform currently also leads to errors.~~

We expect using `jsonProtocol` to improve Prisma Client's startup performance and memory usage **significantly**. This will likely have a more significant impact on applications with larger Prisma schemas.

We would appreciate your feedback on this feature on the following particularly:

1. Does using this preview feature introduce any regressions or problems in your application?
2. If not, how does it influence the performance of your application? Can you share before and after measurements?

For feedback, please comment on this issue.

---

Please share your feedback about the `jsonProtocol` functionality released in [v4.11.0](https://github.com/prisma/prisma/releases/tag/4.11.0) in this issue.

- If you encounter a bug, please [open a bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) in this repo.
- If the feature is working well for you, please share this in a comment below or leave a 👍 on this issue.
- If you have any questions, don't hesitate to ask them in the [Prisma Slack](https://slack.prisma.io/) community

**Comment by n3oney** [comment]
One issue I've found - enum fields don't seem to be selected by default.

For example:

With jsonProtocol
```js
const products = await prisma.db.productInfo.findMany();

/*
{
  "id": "01002ac5-0098-443f-84ac-3294b3169aae",
  "name": "Product one",
  "description": "Product number one",
}
*/
```

Without jsonProtocol
```js
const products = await prisma.db.productInfo.findMany();

/*
{
  "id": "01002ac5-0098-443f-84ac-3294b3169aae",
  "name": "Product one",
  "status": "Enabled",
  "description": "Product number one",
}
*/
```

Adding a `select` to the query and explicitly telling it to select that field seems to fix it.

With jsonProtocol
```js
const products = await prisma.db.productInfo.findMany({
      select: {
        status: true,
        id: true,
        name: true,
        description: true
      }
});


/*
{
  "id": "01002ac5-0098-443f-84ac-3294b3169aae",
  "name": "Product one",
  "status": "Enabled",
  "description": "Product number one",
}
*/
```

**Comment by janpio** [comment]
Two things we discovered:
1. Using Prisma Client with Data Proxy currently sometimes does not lead to good results with `jsonProtocol`. We are investigating.
2. Uploading a Prisma schema with `jsonProtocol` to Prisma Data Platform currently can lead to error messages in Data Browser and Query Console. We will update (but then might be blocked by 1)

**Comment by mmmeff** [comment]
The lack of bug reports is really indicative of the quality of the release. Thanks, Prisma devs!

**Comment by janpio** [comment]
Thanks 😆 The team will love to hear that. 

Note the linked issues #18200 though, we seem to have missed a few details around enums.

And I honestly hope we will get a few "I measured, and this is now x% faster" messages...

**Comment by trentonsnyder** [comment]
@janpio I haven't done a very thorough analysis and my app still needs a lot of tuning but at first glance I'm seeing some pretty decent improvements in performance in Vercel.
I'm using a SQL Server DB with ~8,700 line schema file.

Some observations on one function:

Without "jsonProtocol" I was seeing ~790MB memory usage running upwards of 10 seconds on cold starts.
With "jsonProtocol" I'm seeing ~330MB memory usage and upwards of 6 seconds on cold starts.

**Comment by gmpetrov** [comment]
Hello, I just tried the jsonProtocol option and I noticed that Enum fields are not returned unless select {} is used

```js
 // Works
  const datastores = await prisma.datastore.findMany({
    where: {
      ownerId: session?.user?.id,
    },
    select: {
      // Enum Fields
      visibility: true,
      type: true,

      id: true,
      config: true,
      ownerId: true,
      updatedAt: true,
      description: true,
      name: true,
    },,
  });

// No enum fields returned
  const datastores = await prisma.datastore.findMany({
    where: {
      ownerId: session?.user?.id,
    }
  });


// Schema
model Datastore {
  id String @id @default(cuid())

  name        String
  description String
  type        DatastoreType
  datasources AppDatasource[]
  visibility  DatastoreVisibility @default(private)
  apiKeys     DatastoreApiKey[]

  config Json?

  ownerId String? @map("owner_id")
  owner   User?   @relation(fields: [ownerId], references: [id])

  createdAt DateTime @default(now()) @map("created_at")
  updatedAt DateTime @default(now()) @updatedAt @map("updated_at")

  @@map("data_stores")
}

enum DatasourceType {
  web_page
  text
}

enum DatastoreVisibility {
  public
  private
}

```

**Comment by janpio** [comment]
Thanks @pheuter. Can you maybe pinpoint the query a bit more? Would be amazing if you knew the parameters it uses.

**Comment by semoal** [comment]
In our case after adding `jsonProtocol` to the schema, our testing suite it's failing because `in` it's expected to be always an string[], but without `jsonProtocol` and also the types accepts an Enumerable<T>

```ts
apps/api test:   41 const [count, results] = await prisma.$transaction([
apps/api test: → 42   prisma.market.count({
apps/api test:          select: {
apps/api test:            _count: {
apps/api test:              select: {
apps/api test:                _all: true
apps/api test:              }
apps/api test:            }
apps/api test:          },
apps/api test:          where: {
apps/api test:            marketId: {
apps/api test:              in: "govumfbnfe"
apps/api test:                  ~~~~~~~~~~~~
apps/api test:            }
apps/api test:          }
apps/api test:        })
apps/api test: Argument `in`: Invalid value provided. Expected String[], provided String.
apps/api test:  ❯ src/controllers/v1/markets.controller.ts:42:19
apps/api test:      40|   };
apps/api test:      41|   const [count, results] = await prisma.$transaction([
apps/api test:      42|     prisma.market.count({
apps/api test:        |                   ^
apps/api test:      43|       where: query,
apps/api test:      44|     }),
```

**Comment by michaelhays** [comment]
Just wanted to report that enabling `jsonProtocol` reduced our cold starts on Vercel from ~8.5 seconds to ~4.5 seconds! With no issues found so far. 1,500 line `schema.prisma` file.

Nice work Prisma team :)

**Comment by SevInf** [comment]
@pheuter could you open a separate issue and share more of your schema? At the very least, non-appreviated `Payroll` model and all the models and composites it references. Snippet you've provided is unfortunately not enough to reproduce the issue.

**Comment by SevInf** [comment]
@semoal noted the bug. Until it is fixed, you can replace `markedId: { in: "value" }` part in your test with `marketId: "value"`.

**Comment by yharaskrik** [comment]
I am receiving the following when running with JSON Protocol preview:

Mongo:
```bash
→ 60 await mongo.fundraiser.create(
 $composites: true does not support recursive composite types.

 58 |     });
 59 |
> 60 |     await mongo.fundraiser.create({
```

**Comment by SevInf** [comment]
@yharaskrik please, share full schema and the prisma call so we could reproduce it. Thank you!

**Comment by tavindev** [comment]
Enabling `jsonProtocol` (v4.12.0) reduced RAM usage by ~55% in my case

**Comment by SevInf** [comment]
Separate issue for `$composites: true does not support recursive composite types.` on mongo #18735

**Comment by janpio** [comment]
The `$composites: true does not support recursive composite types.` problem on MongoDB has been fixed @yharaskrik @pheuter . The fix will be included in 4.13.0 next Tuesday, if you are eager to try it out you can use `prisam@dev` and `@prisma/client@dev` right now and report back. Thanks!

**Comment by pheuter** [comment]
Can confirm I no longer see `$composites: true does not support recursive composite types.` errors on my end.

**Comment by janpio** [comment]
[Prisma Data Platform](https://cloud.prisma.io/), especially Data Browser and Query Console, but also Data Proxy should now work properly as well. (Using `PRISMA_ENGINE_PROTOCOL=json` will still lead to errors [as that can conceptually not work], so please use `previewFeatures = ["jsonProtocol"]` for testing with Data Proxy).

**Comment by Ahmadh26** [comment]
Reduced the cold start time by more than half! i'm impressed! great job Prisma team 😄

**Comment by andyjy** [comment]
Working great - with Data Proxy
(after applying temporary workaround for https://github.com/prisma/prisma/issues/18549)

Relatively modest `schema.prisma`: ~1k lines (34 models, 14 enums)

Before:
- time to execute `new PrismaClient()`: ~60-70ms
- Vercel serverless function logs report memory usage: ~205-220MB
- `process.memoryUsage().heapUsed`: 60-70MB

With `jsonProtocol`:
- time to execute `new PrismaClient()`: ~3-12ms (=> ~80-95% improvement)
- Vercel serverless function logs report memory usage: ~160-190MB (=> ~14-22% improvement)
- `process.memoryUsage().heapUsed`: 37-46MB (=> ~35-40% improvement)

**Comment by Lwachira** [comment]
Much progress has been made on the cold start issues, and general data calls are much faster now as well. My app is starting to feel smooth, thank you so much for the work on this, I tried alternatives (pushing to edge, using other db solutions) but keep coming back to prisma.

**Comment by SevInf** [comment]
Hi @shmuelsochet!
That's interesting error. After some internal discussions, we concluded that the JSON protocol correctly reports validation error in this case, and it's the current default, GraphQL protocol that incorrectly turns it into something different from the input and accepts it. More details can be found in #18846, but here is the gist:

Arrays are not valid values for individual elements of `AND`. This code is already rejected by TS compiler and will not pass validation if array is not empty. GraphQL protocol incorrectly encodes empty arrays as empty objects, making it valid for this case, but generally, this is not a correct behavior and should be fixed. To fix your code, ensure that the code that produces the values for `AND` array elements returns objects and not arrays. For the code you provided, that should result in:

```ts
{
  skip: 0,
  take: 25,
  where: {
    AND: [
      {
       AND: [
          {},
          {},
          {},
          undefined
        ]
      },
      {
       column: null
      }
    ]
 }
}
```

**Comment by semoal** [comment]
Latest dev (4.14.0.dev.9) package works like a charm in our side, more than 2000 tests passed correctly without touching anything, on our side we weren't running on any performance bottleneck, but I'm quite sure that this will improve something.

Kudos to the Prisma team, this is the direction that should take any project, focus on performance and stability == high quality.

**Comment by esponges** [comment]
> [Prisma Data Platform](https://cloud.prisma.io/), especially Data Browser and Query Console, but also Data Proxy should now work properly as well. (Using `PRISMA_ENGINE_PROTOCOL=json` will still lead to errors [as that can conceptually not work], so please use `previewFeatures = ["jsonProtocol"]` for testing with Data Proxy).

Where is this `PRISMA_ENGINE_PROTOCOL` flag set? 

My build is failing with

```
error: The preview feature "jsonProtocol" is not known. Expected one of: referentialIntegrity, interactiveTransactions, fullTextSearch, fullTextIndex, tracing, metrics, orderByNulls, filteredRelationCount, fieldReference, postgresqlExtensions, deno, extendedWhereUnique
```

My Prisma version is `4.5.0`.

I'm so excited to test this feature! Thank you

**Comment by alexnault** [comment]
@esponges The Json Protocol is in early preview since 4.11.0. You are using 4.5.0.

**Comment by dbrxnds** [comment]
Hi, we updated to `4.13.0` today and enabled `jsonProtocol`. We can tell a significant difference in performance (Our schema is about 5000 lines, 215 models). But it strangely does break our integration tests, even though dev server & prod build work as expected.

Getting the following error for each test:
```bash
RangeError: Invalid time value
 ❯ serializeArgumentsValue ../../node_modules/@prisma/client-mysql/runtime/library.js:169:3443
 ❯ serializeArgumentsObject ../../node_modules/@prisma/client-mysql/runtime/library.js:169:4063
 ❯ serializeArgumentsValue ../../node_modules/@prisma/client-mysql/runtime/library.js:169:3873
 ❯ serializeArgumentsObject ../../node_modules/@prisma/client-mysql/runtime/library.js:169:4063
 ❯ serializeFieldSelection ../../node_modules/@prisma/client-mysql/runtime/library.js:169:2313
 ❯ serializeJsonQuery ../../node_modules/@prisma/client-mysql/runtime/library.js:169:2216
 ❯ or.createMessage ../../node_modules/@prisma/client-mysql/runtime/library.js:169:5476
 ❯ ../../node_modules/@prisma/client-mysql/runtime/library.js:179:10962
 ❯ runInChildSpan ../../node_modules/@prisma/client-mysql/runtime/library.js:70:25817
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Serialized Error: {
  "clientVersion": "4.13.0",
}
```
Inbefore questions about the `@prisma/client-mysql` path. We do this because we also have a small postgres database which we also use Prisma for. We generate their clients to separate folders this way.

**Comment by Raynos** [comment]
How can I turn this feature on and off at runtime ?

I would love to be able to do

```js
new PrismaClient({
  log: ...,
  datasources: ...,
  jsonProtocol: isStaging ? true : false
})
```

**Comment by SevInf** [comment]
@dbrxnds could you re-run on of the failing tests with `DEBUG=prisma:*` environment variable and post the logs prisma client output? That would greatly help with reproducing it.

**Comment by SevInf** [comment]
@Raynos you can set `PRISMA_ENGINE_PROTOCOL=json` environment variable on your staging deployment. Let me know if that helps.

**Comment by Raynos** [comment]
@SevInf do I need to set that environment variable only for the runtime start of the process or also for `prisma generate` ?

**Comment by dbrxnds** [comment]
Hi @SevInf, I can, but would prefer to do so one-on-one if possible.

**Comment by SevInf** [comment]
@Raynos if you are using DataProxy, env var should be set for both to get all the benefits. If you are not, setting it the start of the process would be enough.

**Comment by SevInf** [comment]
@dbrxnds Sure. You can send the logs and everything you don't want to share publically to schemas@prisma.io and we'll take a look.

**Comment by dbrxnds** [comment]
@SevInf Thank you! Just sent the logs & our current schema file.

**Comment by AlexandroMtzG** [comment]
@SevInf, I have the same error as @dbrxnds on v4.13.0.

- My schema: [100 models](https://gist.github.com/AlexandroMtzG/3093d30b525eda1036b2083f6c176cdb).
- Error: RangeError: Invalid time value.
- Works fine without `previewFeatures = ["jsonProtocol"]`.
- Works when the date value is not null.
- [My logs](https://gist.github.com/AlexandroMtzG/6d4032b97937319537ea6c90d0eea81b) with `DEBUG=prisma:*`, but I don't see any "date" occurrence on the logs before the error (`[createRow] Error RangeError: Invalid time value`).

**Comment by SevInf** [comment]
@Marwen94 with `DEBUG=prisma:*` env var said, there will be a log message with used protocol when you call `new PrismaClient()`:

![Screenshot 2023-04-26 at 16 41 00](https://user-images.githubusercontent.com/603056/234611839-9ea2f7cd-70cf-463b-996d-d7cc2d459878.png)

**Comment by Marwen94** [comment]
@SevInf indeed, thanks!

**Comment by semoal** [comment]
> Hello, we noticed that the following query raises an error after migrating to jsonProtocol : 
> 
> ```
> 
> ctx.prisma.user.findMany({
> 
>       select: { country: true },
> 
>       distinct: 'country',
> 
>       orderBy: { country: 'asc' },
> 
>       where,
> 
>     })
> 
>   },
> 
>  ```
> 
>  the error is : 
> 
>  
> 
>  `Argument distinct : Invalid value provided. Expected UserScalarFieldEnum[], provided string `
> 
> 
> 
> Changing ` distinct: 'country',` to ` distinct: ['country'],` resolves the issue. However, it is weird that typing check does not report this and Typescript compiles successfully. 
> 
>  
> 
>  

This is already fixed for the next version, try prisma version of npm that is tagged under "dev" and the error will be gone!

**Comment by dbrxnds** [comment]
Hi @SevInf, thanks for the research into this issue! 

I have read about this issue before, however the only things we do in regards to dates in the code that is failing, is the Prisma schema having a `@updatedAt`. We don't do anything manually.

**Comment by emroot** [comment]
I'm getting a really obscure bug with vercel and prisma. When jsonProtocol is enabled, I get an Failed to fetch error from apollo on the client when trying to fetch my data from my api. It seems to be happening intermittently after a cold start, when the server is warm it works fine. 
```
[Network error]: TypeError: fetch failed
Unhandled Promise Rejection 	{"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"Error: fetch failed","reason":{"errorType":"Error","errorMessage":"fetch failed","graphQLErrors":[],"clientErrors":[],"networkError":{"errorType":"TypeError","errorMessage":"fetch failed","cause":{"errorType":"HeadersTimeoutError","errorMessage":"Headers Timeout Error","code":"UND_ERR_HEADERS_TIMEOUT","name":"HeadersTimeoutError","message":"Headers Timeout Error","stack":["HeadersTimeoutError: Headers Timeout Error","    at Timeout.onParserTimeout (/var/task/node_modules/next/dist/compiled/undici/index.js:1:60645)","    at Shim.applySegment (/var/task/node_modules/newrelic/lib/shim/shim.js:1297:20)","    at Timeout.wrappedCallback [as _onTimeout] (/var/task/node_modules/newrelic/lib/shim/shim.js:1189:21)","    at listOnTimeout (node:internal/timers:571:11)","    at process.processTimers (node:internal/timers:512:7)"]},"stack":["TypeError: fetch failed","    at Object.fetch (node:internal/deps/undici/undici:11413:11)"]},"message":"fetch failed","stack":["Error: fetch failed","    at new ApolloError (/var/task/node_modules/@apollo/client/errors/errors.cjs:34:28)","    at /var/task/node_modules/@apollo/client/core/core.cjs:1698:19","    at both (/var/task/node_modules/@apollo/client/utilities/utilities.cjs:997:53)","    at /var/task/node_modules/@apollo/client/utilities/utilities.cjs:990:72","    at new Promise (<anonymous>)","    at Object.then (/var/task/node_modules/@apollo/client/utilities/utilities.cjs:990:24)","    at Object.error (/var/task/node_modules/@apollo/client/utilities/utilities.cjs:998:49)","    at notifySubscription (/var/task/node_modules/zen-observable/lib/Observable.js:140:18)","    at onNotify (/var/task/node_modules/zen-observable/lib/Observable.js:179:3)","    at SubscriptionObserver.error (/var/task/node_modules/zen-observable/lib/Observable.js:240:7)"]},"promise":{},"stack":["Runtime.UnhandledPromiseRejection: Error: fetch failed","    at process.<anonymous> (file:///var/runtime/index.mjs:1189:17)","    at process.emit (node:events:525:35)","    at process.wrappedEmit [as emit] (/var/task/node_modules/newrelic/lib/instrumentation/core/globals.js:54:23)","    at emit (node:internal/process/promises:149:20)","    at processPromiseRejections (node:internal/process/promises:283:27)","    at process.processTicksAndRejections (node:internal/process/task_queues:96:32)"]}
Unknown application error occurred
Runtime.Unknown
```
Works fine when jsonProtocol is disabled. Still trying to figure out what's going on but if someone from the prisma team is down to help me debug that would be great.

**Comment by PinkaminaDianePie** [comment]
Still have `$composites: true does not support recursive composite types.` error in 4.13. No idea on how can I debug it, cause I don't even understand what Prisma means by "composite"

**Comment by SevInf** [comment]
@PinkaminaDianePie please, open a new issue and provide a schema and a query you are issuing so we could reproduce it. Thank you!

**Comment by lukasfoisy** [comment]
Hello, I noticed that the data proxy stopped working last night when using the `jsonProtocol`. When removing `previewFeatures = ["jsonProtocol"]` from the Prisma schema, everything works as expected. No changes were made to the schema itself.

- My Vercel builds failed, with the Prisma error being `P5014` and `P5014` on different deployment attempts.
- The Data Browser on `cloud.prisma.io` also didn't work (timed out request)

**Comment by destroyer22719** [comment]
Hello, I have migrated from sequelize and sequelize-typescript to prisma recently, and so far I am loving it for the most part! I recorded a video of me using web scraping with and without `jsonProtocol` and have noticed a noticeable performance increase. The database the code is reading and writing from has ~175MB worth data.

https://user-images.githubusercontent.com/65511595/236702060-8f9a1407-433b-492a-81bb-b620ed895caa.mp4

https://user-images.githubusercontent.com/65511595/236702073-1d9a713b-48f1-4c8c-9f56-63a1aab92b39.mp4

You can see code [here](https://github.com/destroyer22719/FoodFlation/blob/foodflation-v2/scrapers/src/stores/metro.ts) and [here](https://github.com/destroyer22719/FoodFlation/blob/foodflation-v2/scrapers/src/utils/scrapers.ts)

**Comment by janpio** [comment]
@lukasfoisy Can you contact support via https://cloud.prisma.io/? Then we have your project and account id and it is easier to figure out which logs to look at. (Mention that I told you to do so 😆)

**Comment by AlexandroMtzG** [comment]
@SevInf Confirming that is now working 100%. It was indeed an invalid date, thanks! I wish I could say how much it improved, but it feels smoother 👌

**Comment by trevorr** [comment]
We're evaluating `jsonProtocol` with Prisma 4.14 on a large schema (227 models, 7800 non-comment schema lines). We're getting a number of integration test failures with "invalid create invocation". The cause seems to be using the `UncheckedCreateInput` form but Prisma expecting `CreateInput`:

```
Invalid `prisma.contact.create()` invocation:

    {
      data: {
        firstName: "Hayden",
        lastName: "Jast",
        organizationId: "0290f094-69aa-4e87-a7c0-1f208a640286",
    +   organization: {
    +     create: OrganizationCreateWithoutContactsInput | OrganizationUncheckedCreateWithoutContactsInput,
    +     connectOrCreate: OrganizationCreateOrConnectWithoutContactsInput,
    +     connect: OrganizationWhereUniqueInput
    +   }
      }
    }

    Argument `organization` is missing.
```

This type checks fine based on this generated type:

```
  export type ContactUncheckedCreateInput = {
    id?: string
    organizationId: string
    firstName?: string | null
    lastName?: string | null
    ...
```

**Comment by trevorr** [comment]
We're also getting the "Invalid time value" error. While I understand that the JSON protocol is being stricter, the issue is that I can't tell where it's happening because the call stack is not useful. Here's the output from a failed Jest test:

```
    RangeError: Invalid time value
        at Date.toISOString (<anonymous>)

      162 | `)}getCurrentLineLength(){return this.currentLine.length}indentedCurrentLine(){let t=this.currentLine.padStart(this.currentLine.length+2*this.currentIndent);return this.marginSymbol?this.marginSymbol+t.slice(1):t}};var Ja=R(_r());var _e=class{constructor(t,r){this.name=t;this.value=r;this.isRequired=!1}makeRequired(){return this.isRequired=!0,this}write(t){let{colors:{green:r}}=t.context;t.addMarginSymbol(r(this.isRequired?"+":"?")),t.write(r(this.name)),this.isRequired||t.write(r("?")),t.write(r(": ")),typeof this.value=="string"?t.write(r(this.value)):t.write(this.value)}};var fn=e=>e,Va={bold:fn,red:fn,green:fn,dim:fn},Ua={bold:P,red:A,green:S,dim:D},Ct={write(e){e.writeLine(",")}};var ke=class{constructor(t){this.contents=t;this.isUnderlined=!1;this.color=t=>t}underline(){return this.isUnderlined=!0,this}setColor(t){return this.color=t,this}write(t){let r=t.getCurrentLineLength();t.write(this.color(this.contents)),this.isUnderlined&&t.afterNextNewline(()=>{t.write(" ".repeat(r)).writeLine(this.color("~".repeat(this.contents.length)))})}};var Xe=class{constructor(){this.hasError=!1}markAsError(){return this.hasError=!0,this}};var V=class extends Xe{constructor(){super(...arguments);this.fields={};this.suggestions=[]}addField(r){this.fields[r.name]=r}addSuggestion(r){this.suggestions.push(r)}getField(r){return this.fields[r]}getDeepField(r){let[n,...i]=r,o=this.getField(n);if(!o)return;let s=o;for(let a of i){if(!(s.value instanceof V))return;let l=s.value.getField(a);if(!l)return;s=l}return s}getDeepFieldValue(r){return r.length===0?this:this.getDeepField(r)?.value}hasField(r){return Boolean(this.getField(r))}removeAllFields(){this.fields={}}removeField(r){delete this.fields[r]}getFields(){return this.fields}isEmpty(){return Object.keys(this.fields).length===0}getFieldValue(r){return this.getField(r)?.value}getDeepSubSelectionValue(r){let n=this;for(let i of r){if(!(n instanceof V))return;let o=n.getSubSelectionValue(i);if(!o)return;n=o}return n}getDeepSelectionParent(r){let n=this.getSelectionParent();if(!n)return;let i=n;for(let o of r){let s=i.value.getFieldValue(o);if(!s||!(s instanceof V))return;let a=s.getSelectionParent();if(!a)return;i=a}return i}getSelectionParent(){let r=this.getField("select");if(r?.value instanceof V)return{kind:"select",value:r.value};let n=this.getField("include");if(n?.value instanceof V)return{kind:"include",value:n.value}}getSubSelectionValue(r){return this.getSelectionParent()?.value.fields[r].value}getPrintWidth(){let r=Object.values(this.fields);return r.length==0?2:Math.max(...r.map(i=>i.getPrintWidth()))+2}write(r){let n=Object.values(this.fields);if(n.length===0&&this.suggestions.length===0){this.writeEmpty(r);return}this.writeWithContents(r,n)}writeEmpty(r){let n=new ke("{}");this.hasError&&n.setColor(r.context.colors.red).underline(),r.write(n)}writeWithContents(r,n){r.writeLine("{").withIndent(()=>{r.writeJoined(Ct,[...n,...this.suggestions]).newLine()}),r.write("}"),this.hasError&&r.afterNextNewline(()=>{r.writeLine(r.context.colors.red("~".repeat(this.getPrintWidth())))})}};var ee=class extends Xe{constructor(r){super();this.text=r}getPrintWidth(){return this.text.length}write(r){let n=new ke(this.text);this.hasError&&n.underline().setColor(r.context.colors.red),r.write(n)}};var dn=class{constructor(){this.fields=[]}addField(t,r){return this.fields.push({write(n){let{green:i,dim:o}=n.context.colors;n.write(i(o(`${t}: ${r}`))).addMarginSymbol(i(o("+")))}}),this}write(t){let{colors:{green:r}}=t.context;t.writeLine(r("{")).withIndent(()=>{t.writeJoined(Ct,this.fields).newLine()}).write(r("}")).addMarginSymbol(r("+"))}};function Bi(e,t){switch(e.kind){case"IncludeAndSelect":Pf(e,t);break;case"IncludeOnScalar":Mf(e,t);break;case"EmptySelection":Ff(e,t);break;case"UnknownSelectionField":Cf(e,t);break;case"UnknownArgument":Sf(e,t);break;case"UnknownInputField":Af(e,t);break;case"RequiredArgumentMissing":Of(e,t);break;case"InvalidArgumentType":Rf(e,t);break;case"InvalidArgumentValue":$f(e,t);break;case"ValueTooLarge":Df(e,t);break;case"SomeFieldsMissing":kf(e,t);break;case"TooManyFieldsGiven":If(e,t);break;case"Union":Lf(e,t);break;default:throw new Error("not implemented: "+e.kind)}}function Pf(e,t){let r=t.arguments.getDeepSubSelectionValue(e.selectionPath);r&&r instanceof V&&(r.getField("include")?.markAsError(),r.getField("select")?.markAsError()),t.addErrorMessage(n=>`Please ${n.bold("either")} use ${n.green("`include`")} or ${n.green("`select`")}, but ${n.red("not both")} at the same time.`)}function Mf(e,t){let[r,n]=mn(e.selectionPath),i=e.outputType,o=t.arguments.getDeepSelectionParent(r)?.value;if(o&&(o.getField(n)?.markAsError(),i))for(let s of i.fields)s.isRelation&&o.addSuggestion(new _e(s.name,"true"));t.addErrorMessage(s=>{let a=`Invalid scalar field ${s.red(`\`${n}\``)} for ${s.bold("include")} statement`;return i?a+=` on model ${s.bold(i.name)}. ${or(s)}`:a+=".",a+=`
      163 | Note that ${s.bold("include")} statements only accept relation fields.`,a})}function Ff(e,t){let r=e.outputType,n=t.arguments.getDeepSelectionParent(e.selectionPath)?.value,i=n?.isEmpty()??!1;n&&(n.removeAllFields(),Ha(n,r)),t.addErrorMessage(o=>i?`The ${o.red("`select`")} statement for type ${o.bold(r.name)} must not be empty. ${or(o)}`:`The ${o.red("`select`")} statement for type ${o.bold(r.name)} needs ${o.bold("at least one truthy value")}.`)}function Cf(e,t){let[r,n]=mn(e.selectionPath),i=t.arguments.getDeepSelectionParent(r);i&&(i.value.getField(n)?.markAsError(),Ha(i.value,e.outputType)),t.addErrorMessage(o=>{let s=[`Unknown field ${o.red(`\`${n}\``)}`];return i&&s.push(`for ${o.bold(i.kind)} statement`),s.push(`on model ${o.bold(`\`${e.outputType.name}\``)}.`),s.push(or(o)),s.join(" ")})}function Sf(e,t){let r=e.argumentPath[0],n=t.arguments.getDeepSubSelectionValue(e.selectionPath);n instanceof V&&(n.getField(r)?.markAsError(),jf(n,e.arguments)),t.addErrorMessage(i=>Ga(i,r,e.arguments.map(o=>o.name)))}function Af(e,t){let[r,n]=mn(e.argumentPath),i=t.arguments.getDeepSubSelectionValue(e.selectionPath);if(i instanceof V){i.getDeepField(e.argumentPath)?.markAsError();let o=i.getDeepFieldValue(r);o instanceof V&&za(o,e.inputType)}t.addErrorMessage(o=>Ga(o,n,e.inputType.fields.map(s=>s.name)))}function Ga(e,t,r){let n=[`Unknown argument \`${e.red(t)}\`.`],i=Bf(t,r);return i&&n.push(`Did you mean \`${e.green(i)}\`?`),r.length>0&&n.push(or(e)),n.join(" ")}function Of(e,t){let r;t.addErrorMessage(l=>r?.value instanceof ee&&r.value.text==="null"?`Argument \`${l.green(o)}\` must not be ${l.red("null")}.`:`Argument \`${l.green(o)}\` is missing.`);let n=t.arguments.getDeepSubSelectionValue(e.selectionPath);if(!(n instanceof V))return;let[i,o]=mn(e.argumentPath),s=new dn,a=n.getDeepFieldValue(i);if(a instanceof V)if(r=a.getField(o),r&&a.removeField(o),e.inputTypes.length===1&&e.inputTypes[0].kind==="object"){for(let l of e.inputTypes[0].fields)s.addField(l.name,l.typeNames.join(" | "));a.addSuggestion(new _e(o,s).makeRequired())}else{let l=e.inputTypes.map(Ka).join(" | ");a.addSuggestion(new _e(o,l).makeRequired())}}function Ka(e){return e.kind==="list"?`${Ka(e.elementType)}[]`:e.name}function Rf(e,t){let r=e.argument.name,n=t.arguments.getDeepSubSelectionValue(e.selectionPath);n instanceof V&&n.getDeepFieldValue(e.argumentPath)?.markAsError(),t.addErrorMessage(i=>{let o=gn("or",e.argument.typeNames.map(s=>i.green(s)));return`Argument \`${i.bold(r)}\`: Invalid value provided. Expected ${o}, provided ${i.red(e.inferredType)}.`})}function $f(e,t){let r=e.argument.name,n=t.arguments.getDeepSubSelectionValue(e.selectionPath);n instanceof V&&n.getDeepFieldValue(e.argumentPath)?.markAsError(),t.addErrorMessage(i=>{let o=gn("or",e.argument.typeNames.map(a=>i.green(a))),s=[`Invalid value for argument \`${i.bold(r)}\``];return e.underlyingError&&s.push(`: ${e.underlyingError}`),s.push(`. Expected ${o}.`),s.join("")})}function Df(e,t){let r=e.argument.name,n=t.arguments.getDeepSubSelectionValue(e.selectionPath),i;if(n instanceof V){let s=n.getDeepField(e.argumentPath)?.value;s?.markAsError(),s instanceof ee&&(i=s.text)}t.addErrorMessage(o=>{let s=["Unable to fit value"];return i&&s.push(o.red(i)),s.push(`into a 64-bit signed integer for field \`${o.bold(r)}\``),s.join(" ")})}function kf(e,t){let r=e.argumentPath[e.argumentPath.length-1],n=t.arguments.getDeepSubSelectionValue(e.selectionPath);if(n instanceof V){let i=n.getDeepFieldValue(e.argumentPath);i instanceof V&&za(i,e.inputType)}t.addErrorMessage(i=>{let o=[`Argument \`${i.bold(r)}\` of type ${i.bold(e.inputType.name)} needs`];return e.constraints.minFieldCount===1?e.constraints.requiredFields?o.push(`${i.green("at least one of")} ${gn("or",e.constraints.requiredFields.map(s=>`\`${i.bold(s)}\``))} arguments.`):o.push(`${i.green("at least one")} argument.`):o.push(`${i.green(`at least ${e.constraints.minFieldCount}`)} arguments.`),o.push(or(i)),o.join(" ")})}function If(e,t){let r=e.argumentPath[e.argumentPath.length-1],n=t.arguments.getDeepSubSelectionValue(e.selectionPath),i=[];if(n instanceof V){let o=n.getDeepFieldValue(e.argumentPath);o instanceof V&&(o.markAsError(),i=Object.keys(o.getFields()))}t.addErrorMessage(o=>{let s=[`Argument \`${o.bold(r)}\` of type ${o.bold(e.inputType.name)} needs`];return e.constraints.minFieldCount===1&&e.constraints.maxFieldCount==1?s.push(`${o.green("exactly one")} argument,`):e.constraints.maxFieldCount==1?s.push(`${o.green("at most one")} argument,`):s.push(`${o.green(`at most ${e.constraints.maxFieldCount}`)} arguments,`),s.push(`but you provided ${gn("and",i.map(a=>o.red(a)))}. Please choose`),e.constraints.maxFieldCount===1?s.push("one."):s.push(`${e.constraints.maxFieldCount}.`),s.join(" ")})}function Lf(e,t){let r=Wa(e);r?Bi(r,t):t.addErrorMessage(()=>"Unknown error")}function Wa(e){return Nf(e)??_f(e)}function Nf({errors:e}){if(e.length===0||e[0].kind!=="InvalidArgumentType")return;let t={...e[0],argument:{...e[0].argument}};for(let r=1;r<e.length;r++){let n=e[r];if(n.kind!=="InvalidArgumentType"||!Qa(n.selectionPath,t.selectionPath)||!Qa(n.argumentPath,t.argumentPath))return;t.argument.typeNames=t.argument.typeNames.concat(n.argument.typeNames)}return t}function Qa(e,t){if(e.length!==t.length)return!1;for(let r=0;r<e.length;r++)if(e[r]!==t[r])return!1;return!0}function _f(e){return ii(e.errors,t=>{t.kind==="Union"&&(t=Wa(t)??t);let r=0;return Array.isArray(t.selectionPath)&&(r+=t.selectionPath.length),Array.isArray(t.argumentPath)&&(r+=t.argumentPath.length),r})}function Ha(e,t){for(let r of t.fields)e.hasField(r.name)||e.addSuggestion(new _e(r.name,"true"))}function jf(e,t){for(let r of t)e.hasField(r.name)||e.addSuggestion(new _e(r.name,r.typeNames.join(" | ")))}function za(e,t){if(t.kind==="object")for(let r of t.fields)e.hasField(r.name)||e.addSuggestion(new _e(r.name,r.typeNames.join(" | ")))}function mn(e){let t=[...e],r=t.pop();if(!r)throw new Error("unexpected empty path");return[t,r]}function or({green:e}){return`Available options are listed in ${e("green")}.`}function gn(e,t){if(t.length===1)return t[0];let r=[...t],n=r.pop();return`${r.join(", ")} ${e} ${n}`}var qf=3;function Bf(e,t){let r=1/0,n;for(let i of t){let o=(0,Ja.default)(e,i);o>qf||o<r&&(r=o,n=i)}return n}var hn=class extends Xe{constructor(){super(...arguments);this.items=[]}addItem(r){return this.items.push(r),this}getPrintWidth(){return Math.max(...this.items.map(n=>n.getPrintWidth()))+2}write(r){if(this.items.length===0){this.writeEmpty(r);return}this.writeWithItems(r)}writeEmpty(r){let n=new ke("[]");this.hasError&&n.setColor(r.context.colors.red).underline(),r.write(n)}writeWithItems(r){let{colors:n}=r.context;r.writeLine("[").withIndent(()=>r.writeJoined(Ct,this.items).newLine()).write("]"),this.hasError&&r.afterNextNewline(()=>{r.writeLine(n.red("~".repeat(this.getPrintWidth())))})}};var Ya=": ",yn=class{constructor(t,r){this.name=t;this.value=r;this.hasError=!1}markAsError(){this.hasError=!0}getPrintWidth(){return this.name.length+this.value.getPrintWidth()+Ya.length}write(t){let r=new ke(this.name);this.hasError&&r.underline().setColor(t.context.colors.red),t.write(r).write(Ya).write(this.value)}};var Vi=class{constructor(t){this.errorMessages=[];this.arguments=t}write(t){t.write(this.arguments)}addErrorMessage(t){this.errorMessages.push(t)}renderAllMessages(t){return this.errorMessages.map(r=>r(t)).join(`
    > 164 | `)}};function Za(e){return new Vi(Xa(e))}function Xa(e){let t=new V;for(let[r,n]of Object.entries(e)){let i=new yn(r,el(n));t.addField(i)}return t}function el(e){if(typeof e=="string")return new ee(JSON.stringify(e));if(typeof e=="number"||typeof e=="boolean")return new ee(String(e));if(typeof e=="bigint")return new ee(`${e}n`);if(e===null)return new ee("null");if(e===void 0)return new ee("undefined");if(Ge(e))return new ee(`new Prisma.Decimal("${e.toFixed()}")`);if(e instanceof Uint8Array)return Buffer.isBuffer(e)?new ee(`Buffer.alloc(${e.byteLength})`):new ee(`new Uint8Array(${e.byteLength})`);if(e instanceof Date)return new ee(`new Date("${e.toISOString()}")`);if(e instanceof se)return new ee(`Prisma.${e._getName()}`);if(jr(e))return new ee(`prisma.${vt(e.modelName)}.$fields.${e.name}`);if(Array.isArray(e))return Vf(e);if(typeof e=="object")return Xa(e);Ie(e,"Unknown value type")}function Vf(e){let t=new hn;for(let r of e)t.addItem(el(r));return t}function bn({args:e,errors:t,errorFormat:r,callsite:n,originalMethod:i}){let o=Za(e);for(let c of t)Bi(c,o);let s=r==="pretty"?Ua:Va,a=o.renderAllMessages(s),l=new pn(0,{colors:s}).write(o).toString(),u=Re({message:a,callsite:n,originalMethod:i,showColors:r==="pretty",callArguments:l});throw new W(u)}var Uf={findUnique:"findUnique",findUniqueOrThrow:"findUniqueOrThrow",findFirst:"findFirst",findFirstOrThrow:"findFirstOrThrow",findMany:"findMany",count:"aggregate",create:"createOne",createMany:"createMany",update:"updateOne",updateMany:"updateMany",upsert:"upsertOne",delete:"deleteOne",deleteMany:"deleteMany",executeRaw:"executeRaw",queryRaw:"queryRaw",aggregate:"aggregate",groupBy:"groupBy",runCommandRaw:"runCommandRaw",findRaw:"findRaw",aggregateRaw:"aggregateRaw"};function tl({modelName:e,action:t,args:r,runtimeDataModel:n,extensions:i,callsite:o,clientMethod:s,errorFormat:a}){let l=new sr({runtimeDataModel:n,modelName:e,action:t,rootArgs:r,callsite:o,extensions:i,path:[],originalMethod:s,errorFormat:a});return{modelName:e,action:Uf[t],query:Ui(r,l)}}function Ui({select:e,include:t,...r}={},n){return{arguments:nl(r),selection:Qf(e,t,n)}}function Qf(e,t,r){return e&&t&&r.throwValidationError({kind:"IncludeAndSelect",selectionPath:r.getSelectionPath()}),e?Kf(e,r):Jf(r,t)}function Jf(e,t){let r={};return e.model&&!e.isRawAction()&&(r.$composites=!0,r.$scalars=!0),t&&Gf(r,t,e),r}function Gf(e,t,r){for(let[n,i]of Object.entries(t)){let o=r.findField(n);o&&o?.kind!=="object"&&r.throwValidationError({kind:"IncludeOnScalar",selectionPath:r.getSelectionPath().concat(n),outputType:r.getOutputTypeDescription()}),i===!0?e[n]=!0:typeof i=="object"&&(e[n]=Ui(i,r.atField(n)))}}function Kf(e,t){let r={},n=t.getComputedFields(),i=Wr(e,n);for(let[o,s]of Object.entries(i)){let a=t.findField(o);n?.[o]&&!a||(s===!0?r[o]=!0:typeof s=="object"&&(r[o]=Ui(s,t.atField(o))))}return r}function rl(e){if(e===null)return null;if(typeof e=="string"||typeof e=="number"||typeof e=="boolean")return e;if(typeof e=="bigint")return{$type:"BigInt",value:String(e)};if(Hf(e))return{$type:"DateTime",value:e.toISOString()};if(jr(e))return{$type:"FieldRef",value:{_ref:e.name}};if(Array.isArray(e))return Wf(e);if(ArrayBuffer.isView(e))return{$type:"Bytes",value:Buffer.from(e).toString("base64")};if(zf(e))return e.values;if(Ge(e))return{$type:"Decimal",value:e.toFixed()};if(e instanceof se){if(e!==bt.instances[e._getName()])throw new Error("Invalid ObjectEnumValue");return{$type:"Enum",value:e._getName()}}if(typeof e=="object")return nl(e);Ie(e,"Unknown value type")}function nl(e){if(e.$type)return{$type:"Json",value:JSON.stringify(e)};let t={};for(let r in e){let n=e[r];n!==void 0&&(t[r]=rl(n))}return t}function Wf(e){let t=[];for(let r of e)r!==void 0&&t.push(rl(r));return t}function Hf(e){return Object.prototype.toString.call(e)==="[object Date]"}function zf(e){return typeof e=="object"&&e!==null&&e.__prismaRawParameters__===!0}var sr=class{constructor(t){this.params=t;this.params.modelName&&(this.model=this.params.runtimeDataModel.models[this.params.modelName])}throwValidationError(t){bn({errors:[t],originalMethod:this.params.originalMethod,args:this.params.rootArgs??{},callsite:this.params.callsite,errorFormat:this.params.errorFormat})}getSelectionPath(){return this.params.path}getOutputTypeDescription(){if(!(!this.params.modelName||!this.model))return{name:this.params.modelName,fields:this.model.fields.map(t=>({name:t.name,typeName:"boolean",isRelation:t.kind==="object"}))}}isRawAction(){return["executeRaw","queryRaw","runCommandRaw","findRaw","aggregateRaw"].includes(this.params.action)}getComputedFields(){if(!!this.params.modelName)return this.params.extensions.getAllComputedFields(this.params.modelName)}findField(t){return this.model?.fields.find(r=>r.name===t)}atField(t){let r=this.findField(t),n=r?.kind==="object"?r.type:void 0;return new sr({...this.params,modelName:n,path:this.params.path.concat(t)})}};var ar=class{constructor(t,r){this.runtimeDataModel=t;this.errorFormat=r}createMessage(t){let r=tl({...t,runtimeDataModel:this.runtimeDataModel,errorFormat:this.errorFormat});return new wn(r)}createBatch(t){return t.map(r=>r.toEngineQuery())}},Yf={aggregate:!1,aggregateRaw:!1,createMany:!0,createOne:!0,deleteMany:!0,deleteOne:!0,executeRaw:!0,findFirst:!1,findFirstOrThrow:!1,findMany:!1,findRaw:!1,findUnique:!1,findUniqueOrThrow:!1,groupBy:!1,queryRaw:!1,runCommandRaw:!0,updateMany:!0,updateOne:!0,upsertOne:!0},wn=class{constructor(t){this.query=t}isWrite(){return Yf[this.query.action]}getBatchId(){if(this.query.action!=="findUnique")return;let t=[];return this.query.modelName&&t.push(this.query.modelName),this.query.query.arguments&&t.push(Qi(this.query.query.arguments)),t.push(Qi(this.query.query.selection)),t.join("")}toDebugString(){return JSON.stringify(this.query,null,2)}toEngineQuery(){return this.query}deserializeResponse(t,r){if(!t)return t;let n=Object.values(t)[0],i=r.filter(o=>o!=="select"&&o!=="include");return cn(er(n,i))}};function Qi(e){return`(${Object.keys(e).sort().map(r=>{let n=e[r];return typeof n=="object"&&n!==null?`(${r} ${Qi(n)})`:r}).join(" ")})`}var he=class{constructor(t,r){if(t.length-1!==r.length)throw t.length===0?new TypeError("Expected at least 1 string"):new TypeError(`Expected ${t.length} strings to have ${t.length-1} values`);let n=r.reduce((s,a)=>s+(a instanceof he?a.values.length:1),0);this.values=new Array(n),this.strings=new Array(n+1),this.strings[0]=t[0];let i=0,o=0;for(;i<r.length;){let s=r[i++],a=t[i];if(s instanceof he){this.strings[o]+=s.strings[0];let l=0;for(;l<s.values.length;)this.values[o++]=s.values[l++],this.strings[o]=s.strings[l];this.strings[o]+=a}else this.values[o++]=s,this.strings[o]=a}}get text(){let t=1,r=this.strings[0];for(;t<this.strings.length;)r+=`$${t}${this.strings[t++]}`;return r}get sql(){let t=1,r=this.strings[0];for(;t<this.strings.length;)r+=`?${this.strings[t++]}`;return r}inspect(){return{text:this.text,sql:this.sql,values:this.values}}};function il(e,t=",",r="",n=""){if(e.length===0)throw new TypeError("Expected `join([])` to be called with an array of multiple elements, but got an empty array");return new he([r,...Array(e.length-1).fill(t),n],e)}function Ji(e){return new he([e],[])}var ol=Ji("");function Gi(e,...t){return new he(e,t)}var Ki=e=>e.reduce((t,r,n)=>`${t}@P${n}${r}`);function St(e){try{return sl(e,"fast")}catch{return sl(e,"slow")}}function sl(e,t){return JSON.stringify(e.map(r=>Zf(r,t)))}function Zf(e,t){return typeof e=="bigint"?{prisma__type:"bigint",prisma__value:e.toString()}:Xf(e)?{prisma__type:"date",prisma__value:e.toJSON()}:ge.isDecimal(e)?{prisma__type:"decimal",prisma__value:e.toJSON()}:Buffer.isBuffer(e)?{prisma__type:"bytes",prisma__value:e.toString("base64")}:ed(e)||ArrayBuffer.isView(e)?{prisma__type:"bytes",prisma__value:Buffer.from(e).toString("base64")}:typeof e=="object"&&t==="slow"?ll(e):e}function Xf(e){return e instanceof Date?!0:Object.prototype.toString.call(e)==="[object Date]"&&typeof e.toJSON=="function"}function ed(e){return e instanceof ArrayBuffer||e instanceof SharedArrayBuffer?!0:typeof e=="object"&&e!==null?e[Symbol.toStringTag]==="ArrayBuffer"||e[Symbol.toStringTag]==="SharedArrayBuffer":!1}function ll(e){if(typeof e!="object"||e===null)return e;if(typeof e.toJSON=="function")return e.toJSON();if(Array.isArray(e))return e.map(al);let t={};for(let r of Object.keys(e))t[r]=al(e[r]);return t}function al(e){return typeof e=="bigint"?e.toString():ll(e)}var td=/^(\s*alter\s)/i,ul=J("prisma:client");function Wi(e,t,r){if(t.length>0&&td.exec(e))throw new Error(`Running ALTER using ${r} is not supported
          |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               ^
      165 | Using the example below you can still execute your query with Prisma, but please note that it is vulnerable to SQL injection attacks and requires you to take care of input sanitization.
      166 |
      167 | Example:

      at rl (../../shared/src/__generated__/PrismaClient/runtime/library.js:164:3071)
      at nl (../../shared/src/__generated__/PrismaClient/runtime/library.js:164:3660)
      at rl (../../shared/src/__generated__/PrismaClient/runtime/library.js:164:3502)
      at Wf (../../shared/src/__generated__/PrismaClient/runtime/library.js:164:3734)
      at rl (../../shared/src/__generated__/PrismaClient/runtime/library.js:164:3167)
      at nl (../../shared/src/__generated__/PrismaClient/runtime/library.js:164:3660)
      at Ui (../../shared/src/__generated__/PrismaClient/runtime/library.js:164:2096)
      at tl (../../shared/src/__generated__/PrismaClient/runtime/library.js:164:2026)
      at ar.createMessage (../../shared/src/__generated__/PrismaClient/runtime/library.js:164:5007)
```

**Comment by trevorr** [comment]
One final issue: We have an enum array field where we often just pass a single value. The GraphQL protocol and Typescript allow this, but JSON protocol rejects it:

```
Invalid `prisma.contact.createMany()` invocation:

    {
      data: [
        {
          categories: "Investor",
          ...
        }
      ]
    }

    Argument `categories`: Invalid value provided. Expected ContactCreatecategoriesInput or ContactCategory[], provided String.
```

Schema:

```
model Contact {
  ...
  categories ContactCategory[]
```

Generated Typescript:

```ts
  export type ContactCreateManyInput = {
    ...
    categories?: ContactCreatecategoriesInput | Enumerable<ContactCategory>
  }
```

where `Enumerable<T> = T | T[]` and `ContactCategory = "Investor" | ...`.

**Comment by emroot** [comment]
@trevorr sounds like you need to pass an array and not a string
so in that case it should be 
```
{
      data: [
        {
          categories: ["Investor"], // change this line to be an array.
          ...
        }
      ]
    }
```

**Comment by emroot** [comment]
@trevorr I've been running into the same issue, it my case (and for the sake of your example) I was passing organization on creation.

Instead try removing `organization` in the create function to see if that fixes it.

``` typescript
const { organization, ...rest } = payload;
await prisma.contact.create({
  data: {
    ...rest,
  }
})
```

if you wanna create an organization on contact creation then you should use

``` typescript
const { organization, ...rest } = payload;
await prisma.contact.create({
  data: {
    ...rest,
    organization: { 
      create: {
        ...organization
      }
    }
  }
})
```
Something like this, you wanna double check the types but from memory this should do the trick.

Check out the docs: https://www.prisma.io/docs/concepts/components/prisma-schema/relations#relations-in-prisma-client

**Comment by trevorr** [comment]
> @trevorr sounds like you need to pass an array and not a string so in that case it should be

Thanks @emroot. I've done this, but I think it's still a regression. Either JSON protocol is wrong or the generated Typescript is wrong. 😅

**Comment by trevorr** [comment]
> @trevorr I've been running into the same issue, it my case (and for the sake of your example) I was passing organization on creation.

We pass the ID in the unchecked form in a LOT of places, including from programmatically generated queries. It would be a lot of work for us to change it at this point. It sounds more likely that this is a regression rather than an intentional change.

**Comment by janpio** [comment]
Potential remaining problem: https://github.com/prisma/prisma/issues/18735#issuecomment-1532882057

**Comment by yharaskrik** [comment]
When I am testing this out locally I am not seeing a change in memory usage, I am setting the `previewFeatures` to include `jsonProtocol` and then spin up my app with and without the preview feature, in both cases its about 290MB Startup and 238MB Idle, this is within a Nest server and loads 2 different schemas one for mongo (~1600 lines including comments, 40 Models), and one for postgres (~3800 lines and 133 Models).

**Comment by andresribeiro** [comment]
> When I am testing this out locally I am not seeing a change in memory usage, I am setting the `previewFeatures` to include `jsonProtocol` and then spin up my app with and without the preview feature, in both cases its about 290MB Startup and 238MB Idle, this is within a Nest server and loads 2 different schemas one for mongo (~1600 lines including comments, 40 Models), and one for postgres (~3800 lines and 133 Models).

are u running "prisma generate" after changing?

**Comment by yharaskrik** [comment]
> > When I am testing this out locally I am not seeing a change in memory usage, I am setting the `previewFeatures` to include `jsonProtocol` and then spin up my app with and without the preview feature, in both cases its about 290MB Startup and 238MB Idle, this is within a Nest server and loads 2 different schemas one for mongo (~1600 lines including comments, 40 Models), and one for postgres (~3800 lines and 133 Models).
> 
> are u running "prisma generate" after changing?

I did yes. I can try it again though and just triple check. (I actually tried with the env var first but didn't see a change so then tried using the previewFeatures)

I was running the server with `PRISMA_ENGINE_PROTOCOL=json node main.js`

**Comment by SevInf** [comment]
>Potential remaining problem: https://github.com/prisma/prisma/issues/18735#issuecomment-1532882057

#19373

**Comment by SevInf** [comment]
@yharaskrik 
Could you run your script with `DEBUG=prisma:*` env var and ensure JSON protocol is really used during startup? There should be "protocol json" message during `new PrismaClient()` call, like on a screenshot [here](https://github.com/prisma/prisma/issues/18095#issuecomment-1523542451).

**Comment by yharaskrik** [comment]
> @yharaskrik Could you run your script with `DEBUG=prisma:*` env var and ensure JSON protocol is really used during startup? There should be "protocol json" message during `new PrismaClient()` call, like on a screenshot [here](https://github.com/prisma/prisma/issues/18095#issuecomment-1523542451).

Ok I tried again and checked the protocol in the debug messages (measurements in MB)

JSON
237-295

Graphql
245-302

**Comment by yharaskrik** [comment]
> > @yharaskrik Could you run your script with `DEBUG=prisma:*` env var and ensure JSON protocol is really used during startup? There should be "protocol json" message during `new PrismaClient()` call, like on a screenshot [here](https://github.com/prisma/prisma/issues/18095#issuecomment-1523542451).
> 
> Ok I tried again and checked the protocol in the debug messages (measurements in MB)
> 
> JSON 237-295
> 
> Graphql 245-302

Did some additional measuring with and without Prisma with the API idling:

```
Without prisma Idle at: 161MB
With prisma (json) idle at: 225MB
With prisma (graphql) idle at: 227MB
```

So Prisma adds about 65MB regardless of protocol.

**Comment by giva9712** [comment]
> Hello, I have migrated from sequelize and sequelize-typescript to prisma recently, and so far I am loving it for the most part! I recorded a video of me using web scraping with and without `jsonProtocol` and have noticed a noticeable performance increase. The database the code is reading and writing from has ~175MB worth data.
> 
>  foodflation_scraper_jsonDecoder-2023-05-07_16.49.00.mp4 
>  foodflation_scraper_jsonDecoder-2023-05-07_16.23.18.mp4 
> You can see code [here](https://github.com/destroyer22719/FoodFlation/blob/foodflation-v2/scrapers/src/stores/metro.ts) and [here](https://github.com/destroyer22719/FoodFlation/blob/foodflation-v2/scrapers/src/utils/scrapers.ts)

@destroyer22719 

In your two videos, they are almost equal, even I didn't see a 2, 3 seconds speed difference

**Comment by perenstrom** [comment]
I've tested a bit now with an example dataset I found, quite few models, quite few rows in the schema. But I went from ~4500ms to ~500ms in cold start times on Vercel serverless (NextJS API functions). So it cut about 90% of the time. Great job, will continue to explore!

**Comment by immjs** [comment]
> One final issue: We have an enum array field where we often just pass a single value. The GraphQL protocol and Typescript allow this, but JSON protocol rejects it:
> 
> ```
> Invalid `prisma.contact.createMany()` invocation:
> 
>     {
>       data: [
>         {
>           categories: "Investor",
>           ...
>         }
>       ]
>     }
> 
>     Argument `categories`: Invalid value provided. Expected ContactCreatecategoriesInput or ContactCategory[], provided String.
> ```
> 
> Schema:
> 
> ```
> model Contact {
>   ...
>   categories ContactCategory[]
> ```
> 
> Generated Typescript:
> 
> ```ts
>   export type ContactCreateManyInput = {
>     ...
>     categories?: ContactCreatecategoriesInput | Enumerable<ContactCategory>
>   }
> ```
> 
> where `Enumerable<T> = T | T[]` and `ContactCategory = "Investor" | ...`.

Did you find a solution? I am experiencing the same issue

**Comment by immjs** [comment]
> We're evaluating `jsonProtocol` with Prisma 4.14 on a large schema (227 models, 7800 non-comment schema lines). We're getting a number of integration test failures with "invalid create invocation". The cause seems to be using the `UncheckedCreateInput` form but Prisma expecting `CreateInput`:
> 
> ```
> Invalid `prisma.contact.create()` invocation:
> 
>     {
>       data: {
>         firstName: "Hayden",
>         lastName: "Jast",
>         organizationId: "0290f094-69aa-4e87-a7c0-1f208a640286",
>     +   organization: {
>     +     create: OrganizationCreateWithoutContactsInput | OrganizationUncheckedCreateWithoutContactsInput,
>     +     connectOrCreate: OrganizationCreateOrConnectWithoutContactsInput,
>     +     connect: OrganizationWhereUniqueInput
>     +   }
>       }
>     }
> 
>     Argument `organization` is missing.
> ```
> 
> This type checks fine based on this generated type:
> 
> ```
>   export type ContactUncheckedCreateInput = {
>     id?: string
>     organizationId: string
>     firstName?: string | null
>     lastName?: string | null
>     ...
> ```

(Sorry, quote replied to the wrong message)

Did you figure out a solution to this one? I am experiencing the same issue

```
Invalid `prisma.convo.create()` invocation:
{
  data: {
    subject: "a",
    interlocutors: "[\"----\"]",
    folderId: "cliebhdly0002me088n3nxe3c",
    latest: new Date("2023-06-21T16:09:59.874Z"),
+   folder: {
+     create: FolderCreateWithoutConvosInput | FolderUncheckedCreateWithoutConvosInput,
+     connectOrCreate: FolderCreateOrConnectWithoutConvosInput,
+     connect: FolderWhereUniqueInput
+   }
  }
}
Argument `folder` is missing.
```

**Comment by johnjcsmith** [comment]
Hello, 

We have been testing `jsonProtocol` with Prisma `4.14.1` and have seen some instances of services total memory usage dropping by 50%. Amazing work!


We have run into one "issue":
When filtering on JSON object properties, we had some Prisma code which looks like the following:
```
await prisma.foo.findMany({
  where: {
    nestedJsonObject: {
      path: 'bar',
      equals: 'baz',
    },
  },
})
```

Which started to emit the following error after enabling `jsonProtocol`:
```
Argument `path`: Invalid value provided. Expected String[], provided String.
```

We have since updated `path` to a `string[]` as per [the docs](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields/working-with-json-fields#filter-on-object-property).

My understanding is that this is likely the _intended_ behaviour. Just flagging that it does appear to have become stricter with the new protocol.

**Comment by janpio** [comment]
Correct @johnjcsmith, this is one of multiple changes to the "array shortcuts" that we did to clean up our APIs. We will introduce these as breaking changes in Prisma 5 when we make `jsonProtocol` GA and available by default. Good to hear the change was a simple one for you. Thanks!

**Comment by Jolg42** [comment]
Thank you everyone for the feedback 🙌🏼 

We’re thrilled to announce that the `jsonProtocol` Preview feature is now Generally Available. You can now remove the Preview feature flag from your schema after upgrading. We made the JSON-based wire protocol the **default protocol** used for communication between Prisma Client and the query engine.

We introduced this feature in version [4.11.0](https://github.com/prisma/prisma/releases/tag/4.11.0) to improve Prisma’s performance. Previously, Prisma used a GraphQL-like protocol to communicate between Prisma Client and the query engine. Applications with larger schemas had higher CPU and memory consumption compared to smaller schemas which created a performance bottleneck.

The JSON-based wire protocol improves efficiency when Prisma Client is communicating with the query engine.

If you want to learn more about the performance improvements that are bundled with Prisma 5, including the JSON-based protocol, we recommend giving [Prisma 5: Faster by Default](https://www.prisma.io/blog/prisma-5-f66prwkjx72s) a read.

See [Prisma 5.0.0 release notes.](https://github.com/prisma/prisma/releases/tag/5.0.0)

**Comment by ayvanov** [comment]
> If you want to learn more about the performance improvements that are bundled with Prisma 5, including the JSON-based protocol, we recommend giving [Prisma 5: Faster by Default](https://www.prisma.io/blog/prisma-5-f66prwkjx72s) a read.

Link is 404?

**Comment by Jolg42** [comment]
@ayvanov Sorry about that, stay tuned, the blog post is not published yet but should be coming very soon!

**Comment by Jolg42** [comment]
@ayvanov Here it is https://www.prisma.io/blog/prisma-5-f66prwkjx72s ✍🏼
